### PR TITLE
Do not use Numba objmode for supported advanced indexing operations

### PR DESCRIPTION
**Pull request by ricardoV94**
## Description

Several cases of advanced indexing are supported by Numba these days. This PR makes sure we don't skip on those!

## Related Issue

- [ ] Closes #
- [x] Related to #772 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by ricardoV94** [comment]
@kc611 sorry for tagging if your review, fat fingers on my side! Do keep the cool work on extending advanced indexing on the Numba side!!!

File: pytensor/link/numba/dispatch/basic.py [hunk 1]
```diff
@@ -572,48 +560,43 @@ def {fn_name}({", ".join(input_names)}):
...
-            UserWarning,
-        )
-
-    incsubtensor_def_src = create_index_func(node, objmode=objmode)
-
-    global_env = {"np": np}
-    if objmode:
-        global_env["objmode"] = numba.objmode
+@numba_funcify.register(AdvancedSubtensor)
+@numba_funcify.register(AdvancedIncSubtensor)
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/basic.py:577]: Should add a check for broadcasted y values in inc operations, and for the case reported in https://github.com/numba/numba/issues/9573

**Comment by kc611** [comment]
> @kc611 sorry for tagging if your review, fat fingers on my side! Do keep the cool work on extending advanced indexing on the Numba side!!!

No worries! :slightly_smiling_face: I took a quick look at this PR and I'd say if you guys are using string-based function creation anyways then it's completely possible to support the full set of advanced indexing cases by simply creating the required loops within the string-based functions.

**Comment by ricardoV94** [comment]
> > @kc611 sorry for tagging if your review, fat fingers on my side! Do keep the cool work on extending advanced indexing on the Numba side!!!
> 
> No worries! 🙂 I took a quick look at this PR and I'd say if you guys are using string-based function creation anyways then it's completely possible to support the full set of advanced indexing cases by simply creating the required loops within the string-based functions.

Yeah, but we are not doing that yet. We are just creating dumb stuff like `x[indices] = y`. We'll come to that if we need it, or perhaps Numba will support it first :D  

Thanks for checking it out!

**Comment by ricardoV94** [comment]
This should be ready for review. After this we can start adding some simple non-supported cases like boolean indexing with indices.shape == x.shape, and some multiple vector index cases

**Review by jessegrabowski — APPROVED** [review]
Looks good to me. The duplicate indices thing isn't really clear to me. Also curious if the lines codecov is complaining about are unreachable, and, if so, whether we can just get rid of them.

File: pytensor/link/numba/dispatch/subtensor.py [hunk 2]
```diff
@@ -74,84 +74,57 @@ def convert_indices(indices, entry):
...
+        idx.type.ndim
+        for idx in idxs
+        if (isinstance(idx.type, TensorType) and idx.type.ndim > 0)
+    ]
+
+    if (
+        # Numba does not support indexes with more than one dimension
+        # Nor multiple vector indexes
+        (len(adv_idxs_dims) > 1 or adv_idxs_dims[0] > 1)
+        # The default index implementation does not handle duplicate indices correctly
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/link/numba/dispatch/subtensor.py:118]: It means the code returned by `numba_funcify_default_subtensor`. Shouldn't be a problem as long as we don't try to use it when it's not valid

**Pull request merged**